src/reducers/custom-wildcard-field.ts

```diff
@@ -0,0 +1,92 @@
+
+import {Action} from '../actions';
+import {CUSTOM_WILDCARD_ADD, CUSTOM_WILDCARD_ADD_FIELD, CUSTOM_WILDCARD_MODIFY_TITLE,
+        CUSTOM_WILDCARD_REMOVE, CUSTOM_WILDCARD_REMOVE_FIELD} from '../actions/custom-wildcard-field';
+import {CustomWildcardField} from '../models/custom-wildcard-field';
+import {insertItemToArray, modifyItemInArray, removeItemFromArray} from './util';
+
+function modifyFieldsProperty(fields: string[]) {
+  return (customWildcardField: CustomWildcardField) => {
+    return {
+      ...customWildcardField,
+      fields
+    };
+  };
+}
+
+export function customWildcardFieldReducer(
+  customWildcardFields: Readonly<CustomWildcardField[]> = [],
+  action: Action
+): CustomWildcardField[] {
+  switch (action.type) {
+    case CUSTOM_WILDCARD_ADD: {
+      const {fields, type} = action.payload;
+
+      let index = action.payload.index;
+      if (!index) {
+        index = customWildcardFields.length;
+      }
+
+      return insertItemToArray(customWildcardFields, index, {
+        fields,
+        type,
+        title: null
+      });
+    }
+
+    case CUSTOM_WILDCARD_REMOVE: {
+      const {index} = action.payload;
+      return removeItemFromArray(customWildcardFields, index).array;
+    }
+
+    case CUSTOM_WILDCARD_ADD_FIELD: {
+      const {index, fields} = action.payload;
+
+      const originalFields: string[] = customWildcardFields[index].fields;
+      const originalFieldsIndex = originalFields.reduce((dict, field) => {
+        dict[field] = true;
+        return dict;
+      }, {});
+
+      let updatedFields: string[];
+
+      if (fields.length > 1) {
+        updatedFields = [].concat(originalFields);
+        for (const field of fields) {
+          if (!originalFieldsIndex[field]) {
+            updatedFields.push(field);
+          }
+        }
+      } else {
+        if (!originalFieldsIndex[fields[0]]) {
+          updatedFields = originalFields.concat(fields);
+        }
+      }
+
+      return modifyItemInArray(customWildcardFields, index, modifyFieldsProperty(updatedFields));
+    }
+
+    case CUSTOM_WILDCARD_REMOVE_FIELD: {
+      const {index, field} = action.payload;
+      const originalFields: string[] = customWildcardFields[index].fields;
+      const updatedFields = originalFields.filter(originalField => originalField !== field);
+
+      return modifyItemInArray(customWildcardFields, index, modifyFieldsProperty(updatedFields));
```

Comment:
Similar problem with immutability here.

Reply:
N/A - a remove action is triggered because the user deletes an existing field -- array.filter() will always return a new array